proposals/2023/05-26_vela_ui_rewrite.md

```diff
@@ -0,0 +1,217 @@
+
+React, like most frontend tooling, will enter its halcyon days and the industry will move to new tooling. There is no such thing as future-proof. However, in an effort to be future-resistant we are specifically making choices that will be easy to change around build tooling. We are not entirely leveraging the React core team's ecosystem (Next, CRA, Jest), and instead using a broader community ecosystem (Vite).
+
+</details>
+
+<details>
+<summary><b>Can we achieve the same and better results than Elm with this new approach?</b></summary>
+
+Elm offered a bundled developer experience for webapps. The React-centric ecosystem is much more fragmented, but it lets you pick from a variety of options that are easier to plug in and to swap as needs change.
+
+</details>
+
+We have considered these functional changes:
+
+- new/updated endpoints or url paths
+  - re-evaluating the `/-/secrets` path
+  - re-evaluating paths that shadow potentially valid Organization, Repository, or Secret names
+- new/updated configuration variables (environment, flags, files, etc.)
+  - environment variables for features (e.g. default pull request settings, auto-linkifying in logs, etc)
+
+From an overall perspective, the goal is to:
+
+- Replicate the current user interface (look'n'feel; design) in the technology stack listed above as closely as possible
+- Replicate the current functionality in the technology stack listed above as closely as possible
+
+Consider this redesign and rewrite a `1-to-1.3̅3̅3̅` port rather than a `1-to-1` port.
+
+Further enhancements can and will be made after.
+
+## Implementation
+
+<!--
+This section is intended to explain how the solution will be implemented for the proposal.
+
+NOTE: If there are no current plans for implementation, please leave this section blank.
+-->
+
+- Since the UI is a different repo and there are no co-dependency's the new UI should be a drop-in replacement and in theory should be able to run in parallel to the existing version if necessary.
+
+**Please briefly answer the following questions:**
+
+1. Is this something you plan to implement yourself?
+
+Yes. There is a small team working on this (see below; hackathon activity).
+
+2. What's the estimated time to completion?
+
+30-60 days
+
+1. During the 2023 Summer Vela Hackathon, we explored this space and within roughly 62 hours we have ported a majority of the existing look'n'feel and functionality
+2. While a majority of read-centric screens exist, there are write-centric screens that need implementation
+3. There is still plenty of polishing, documentation and cleanup activities thereafter
+
+**Gists, issues, pull requests, etc**
+
+Please refer to the ongoing work in [`go-vela/ui-hackathon`](https://github.com/go-vela/ui-hackathon) repo for more details.
+
+## Questions
+
+**Please list any questions you may have:**
+
```

Comment:
Can you address "maintenance" after this would be implemented? Who is responsible for maintaining all the dependencies? What commitments should be made for breaking changes? Security updates are an important factor at Target that require 30 or 90 day updates - so would those all fall to Target? It would be nice to see more details around who is responsible after everything is released.

Reply:
- Can you address "maintenance" after this would be implemented?
  - With renovate and some documentation, maintenance tasks should be fairly low effort
- Who is responsible for maintaining all the dependencies?
  - Who maintained this for Elm?
  - Related to the topic of contributors, I'd hope there's more than just a replacement of primary contributors. Meanwhile, renovate should certainly pickup a majority of changes
- What commitments should be made for breaking changes?
  - What commitments exist today?
  - Breaking changes for user interface are different than breaking changes for api
  - AFAIK, this is pre 1.0 from some perspective
- Security updates are an important factor at Target that require 30 or 90 day updates - so would those all fall to Target?
  - I would like to learn more about this. Have there been many major security incidents with this user interface in the past? Have regular patch versions with upgrades been rolled out from the open source repo? Excluding the nginx host and other server side bits?
  - The surface area does change from Elm to a more JS-centric ecosystem
  - Again, with renovate and the ease of updating dependencies, it should not be overwhelming regardless of which team or community members are engaging with it

Reply:
Sounds like all the maintenance would fall to Target. There are a couple ways to look at Elm but without any updates, there aren't any we need to apply - which has eliminated our maintenance effort.

Reviewing the React release cycles, it looks like there's a major release once a year with a couple updates in the following months. Depending how our Security team classifies these, each update would require us to take action on either a 30 day or 90 day timeline. From experience, the 30 day cycles are very rough on our team becaus we need to drop everything we're doing and retool for the update. Enough of those back-to-back (like React had in 2022) and we fall behind on feature development because the last 6 months are spent just updating React. Breaking changes (like the ones called out in 18.0.0) extend our testing cycles. At at time when we're trying to be more predictable in our rollouts (1.5-2 months), something like 18.0.0 provides no value but takes up a lot of time.

https://github.com/facebook/react/releases 

Because Elm doesn't require any maintenance today, React will force us to conduct maintenance. While this isn't monthly, it still appear to be disruptive to our schedules.